lib/fs/basicfs_copy_range_copyfilerange.go

```diff
@@ -0,0 +1,44 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// +build linux
+
+package fs
+
+import (
+	"io"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+func init() {
+	registerCopyRangeImplementation(copyRangeImplementation{
+		name: "copy_file_range",
+		impl: copyRangeCopyFileRange,
+	})
+}
+
+func copyRangeCopyFileRange(src, dst basicFile, srcOffset, dstOffset, size int64) error {
+	for size > 0 {
+		// From MAN page:
+		//
+		// If off_in is not NULL, then off_in must point to a buffer that
+		// specifies the starting offset where bytes from fd_in will be read.
+		// 	The file offset of fd_in is not changed, but off_in is adjusted
+		// appropriately.
+		n, err := unix.CopyFileRange(int(src.Fd()), &srcOffset, int(dst.Fd()), &dstOffset, int(size), 0)
+		if n == 0 && err == nil {
+			return io.ErrUnexpectedEOF
+		}
+		if err != nil && err != syscall.EAGAIN {
+			return err
+		}
+		dstOffset += int64(n)
```

Comment:
> The file offset of fd_in is not changed, but off_in is adjusted appropriately

I guess you take this to mean that srcOffset is automatically increased, even in the case of error, but dstOffset is not? Are you sure about that, on both accounts?

Reply:
🤷‍♂  useful content in the man page:
```
The following semantics apply for off_in, and similar statements
       apply to off_out:
```

how simillar exactly? :suspect:

Reply:
Apparently you are right, assertion of the dstOffset after write changing verifies that.

I can't really write a test for it, as, I can't force unix.CopyRange to return prematurely to catch writes happening in the wrong place.